### PR TITLE
Add jsonb purl column to qualified_purl.

### PR DESCRIPTION
**Pull request by JimFuller-RedHat**
Adds a jsonb purl column to qualified_purls, fixing #771.

Adjusts **get_purl** pg func to take advantage of this column (which makes for faster analysis graph processing).

### CODE REVIEW

File: entity/src/qualified_purl.rs [hunk 1]
```diff
@@ -1,5 +1,141 @@
...
+use std::fmt::{Display, Formatter};
+use trustify_common::purl::Purl;
+
+// TODO: some day we might 'collapse' all the purl structs to this one, that day is not today.
+//
+// Instead of using composition or directly relating to any existing purl struct we will implement
+// to & from impl and in places we may invalidate DRY principle by copying across required code blocks.
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, FromJsonQueryResult)]
+pub struct CanonicalPurl {
```
Comment (bobmcwhirter) [inline entity/src/qualified_purl.rs:15]: I suspect this could replace (or be replaced by?) `trustify_common::purl::Purl`.

Comment (bobmcwhirter) [reply entity/src/qualified_purl.rs:15]: (non-blocking for this PR; iterate!)

File: entity/src/qualified_purl.rs [hunk 2]
```diff
@@ -1,5 +1,141 @@
...
+    pub ty: String,
+    pub namespace: Option<String>,
+    pub name: String,
+    pub version: Option<String>,
+    pub qualifiers: BTreeMap<String, String>,
+}
+
+// inspired by common/src/purl.rs#135 but avoids any temporary strings for speed
```
Comment (bobmcwhirter) [inline entity/src/qualified_purl.rs:23]: nice!

File: migration/src/lib.rs [hunk 3]
```diff
@@ -160,6 +161,7 @@ impl MigratorTrait for Migrator {
             Box::new(m0000605_create_source_document::Migration),
             Box::new(m0000610_improve_version_cmp_fns::Migration),
             Box::new(m0000620_parallel_unsafe_pg_fns::Migration),
+            Box::new(m0000630_alter_qualified_purl_purl_column::Migration),
```
Comment (bobmcwhirter) [inline migration/src/lib.rs:164]: Coordinate with @dejanb as I think he's also using `630` migration number.

File: migration/src/m0000630_alter_qualified_purl_purl_column.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                CREATE INDEX QualifiedPurlNameJsonGistIdx ON qualified_purl USING GIST ((purl ->> 'name') gist_trgm_ops);
+                CREATE INDEX QualifiedPurlVersionJsonGistIdx ON qualified_purl USING GIST ((purl ->> 'version') gist_trgm_ops);
+                CREATE INDEX QualifiedPurlNamespaceJsonGistIdx ON qualified_purl USING GIST ((purl ->> 'namespace') gist_trgm_ops);
+                CREATE INDEX QualifiedPurlTypeJsonGistIdx ON qualified_purl USING GIST ((purl ->> 'ty') gist_trgm_ops);
+                "#,
+            )
+            .await?;
+
+        //data migration
+        let _data_migration_result = manager
```
Comment (bobmcwhirter) [inline migration/src/m0000630_alter_qualified_purl_purl_column.rs:33]: I think the `?` at the end counts as "using" the result, so you can avoid the assignment to a `_var`.

File: migration/src/m0000630_alter_qualified_purl_purl_column.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            )
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[allow(clippy::enum_variant_names)]
+#[derive(DeriveIden)]
+pub enum Indexes {
```
Comment (bobmcwhirter) [inline migration/src/m0000630_alter_qualified_purl_purl_column.rs:98]: elide 'pub'

File: modules/fundamental/src/sbom/service/sbom.rs [hunk 6]
```diff
@@ -494,6 +496,18 @@ async fn package_from_row(
 
     for purl in row.purls {
         if let Ok(dto) = serde_json::from_value::<PurlDto>(purl) {
+            let cp = CanonicalPurl {
```
Comment (bobmcwhirter) [inline modules/fundamental/src/sbom/service/sbom.rs:499]: oh, yeah, I forgot we _also_ have `PurlDto`.  Yo dawg, I heard you liked pURLs, so I put a pURL in your pURL so you can pURL while you pURL.

_sigh_

**Review by bobmcwhirter — APPROVED** [review]
No blockers, just comments. LGTM.

File: migration/src/lib.rs (hunk 3 above)
Comment (JimFuller-RedHat) [reply migration/src/lib.rs:164]: @dejanb I will use 640 ;)

File: modules/fundamental/src/sbom/service/sbom.rs (hunk 6 above)
Comment (JimFuller-RedHat) [reply modules/fundamental/src/sbom/service/sbom.rs:499]: one purl to rule them all, one purl to bind them ....

**Pull request merged**